pkg/background/generate/generate.go

```diff
@@ -372,182 +372,215 @@ func (c *GenerateController) ApplyGeneratePolicy(log logr.Logger, policyContext
 }
 
 func getResourceInfo(object map[string]interface{}) (kind, name, namespace, apiversion string, err error) {
-	if kind, _, err = unstructured.NestedString(object, "kind"); err != nil {
-		return "", "", "", "", err
-	}
+	if _, found, err := unstructured.NestedStringSlice(object, "kinds"); !found {
+		if kind, _, err = unstructured.NestedString(object, "kind"); err != nil {
+			return "", "", "", "", err
+		}
 
-	if name, _, err = unstructured.NestedString(object, "name"); err != nil {
-		return "", "", "", "", err
-	}
+		if name, _, err = unstructured.NestedString(object, "name"); err != nil {
+			return "", "", "", "", err
+		}
 
+		if apiversion, _, err = unstructured.NestedString(object, "apiVersion"); err != nil {
+			return "", "", "", "", err
+		}
+	}
 	if namespace, _, err = unstructured.NestedString(object, "namespace"); err != nil {
 		return "", "", "", "", err
 	}
+	return
+}
 
-	if apiversion, _, err = unstructured.NestedString(object, "apiVersion"); err != nil {
-		return "", "", "", "", err
+func getResourceInfoFromRule(rule kyvernov1.Rule) (kind, name, namespace, apiversion string, err error) {
+	if len(rule.Generation.CloneList.Kinds) == 0 {
+		if kind = rule.Generation.Kind; kind == "" {
+			return "", "", "", "", fmt.Errorf("%s", "kind can not be empty")
+		}
+		if name = rule.Generation.Name; name == "" {
+			return "", "", "", "", fmt.Errorf("%s", "name can not be empty")
+		}
+		if apiversion = rule.Generation.APIVersion; apiversion == "" {
+			return "", "", "", "", fmt.Errorf("%s", "apiVersion can not be empty")
+		}
+	}
+	if namespace = rule.Generation.Namespace; namespace == "" {
+		return "", "", "", "", fmt.Errorf("%s", "namespace can not be empty")
```

Comment:
Why is this check out of the `if len(rule.Generation.CloneList.Kinds) == 0{}` condition?

Reply:
`generate.namespace` is required field even with kinds list , how do we know which is target namespace ?

Reply:
got it, keeping it consistent to support backward compatibility, made it optional in case of cluster scope resource.